dask/dataframe/core.py

```diff
@@ -2330,6 +2335,236 @@ def quantile(df, q):
+    if num_fixed + num_random + 5 >= chunk_length:
+        return np.linspace(0, 100, chunk_length + 1)
+
+    if not isinstance(random_state, np.random.RandomState):
+        random_state = np.random.RandomState(random_state)
+
+    q_fixed = np.linspace(0, 100, num_fixed)
+    q_random = random_state.rand(num_random) * 100
+    q_edges = [0.5 / num_fixed, 1 - 0.5 / num_fixed]
+    qs = np.concatenate([q_fixed, q_random, q_edges])
+    qs.sort()
+    return qs
+
+
+def tree_width(N, to_binary=False):
+    """Generate tree width suitable for ``merge_sorted`` given N inputs
+
+    In theory, this is designed so all tasks are of comparable effort.
+    """
+    if N < 32:
+        group_size = 2
+    else:
+        group_size = int(math.log(N))
+    num_groups = N // group_size
+    if to_binary or num_groups < 16:
+        return 2**int(math.log(float(N) / group_size, 2))
+    else:
+        return num_groups
+
+
+def tree_groups(N, num_groups):
+    """Split an integer N into evenly sized and spaced groups.
+
+    >>> tree_groups(16, 6)
+    [3, 2, 3, 3, 2, 3]
+    """
+    # Bresenham, you so smooth!
+    group_size = N // num_groups
+    dx = num_groups
+    dy = N - group_size * num_groups
+    D = 2*dy - dx
+    rv = []
+    for _ in range(num_groups):
+        if D < 0:
+            rv.append(group_size)
+        else:
+            rv.append(group_size + 1)
+            D -= 2*dx
+        D += 2*dy
+    return rv
+
+
+def create_merge_tree(func, keys, token):
+    """Create a task tree that merges all the keys with a reduction function.
+
+    Parameters
+    ----------
+    func: callable
+        Reduction function that accepts a single list of values to reduce.
+    keys: iterable
+        Keys to reduce from the source dask graph.
+    token: object
+        Included in each key of the returned dict.
+
+    This creates a k-ary tree where k depends on the current level and is
+    greater the further away a node is from the root node.  This reduces the
+    total number of nodes (thereby reducing scheduler overhead), but still
+    has beneficial properties of trees.
+
+    For reasonable numbers of keys, N < 1e5, the total number of nodes in the
+    tree is approximately N**0.78.
+    """
+    level = 0
+    prev_width = len(keys)
+    prev_keys = iter(keys)
+    rv = {}
+    while prev_width > 1:
+        width = tree_width(prev_width)
+        groups = tree_groups(prev_width, width)
+        keys = [(token, level, i) for i in range(width)]
+        rv.update((key, (func, list(take(num, prev_keys))))
+                   for num, key in zip(groups, keys))
+        prev_width = width
+        prev_keys = iter(keys)
+        level += 1
+    return rv
+
+
+def _merge_sorted(items):
+    return list(merge_sorted(*items))
+
+
+def _prepare_percentile_merge(qs, vals, length):
+    if length == 0:
+        return []
+    diff = np.ediff1d(qs, 0.0, 0.0)
+    weights = 0.5 * length * (diff[1:] + diff[:-1])
+    return list(zip(vals, weights))
```

Comment:
After a brief look it's not clear to me what this function does.  Docstring appreciated